crypto/shamir.c

```diff
@@ -0,0 +1,329 @@
+  r[4] ^= a2[7] & b[5];
+  r[5] ^= a2[0] & b[5];
+  r[6] ^= a2[1] & b[5];
+  r[7] ^= a2[2] & b[5];
+  a2[3] ^= a2[2]; /* reduce */
+  a2[5] ^= a2[2];
+  a2[6] ^= a2[2];
+
+  r[0] ^= a2[2] & b[6]; /* add */
+  r[1] ^= a2[3] & b[6];
+  r[2] ^= a2[4] & b[6];
+  r[3] ^= a2[5] & b[6];
+  r[4] ^= a2[6] & b[6];
+  r[5] ^= a2[7] & b[6];
+  r[6] ^= a2[0] & b[6];
+  r[7] ^= a2[1] & b[6];
+  a2[2] ^= a2[1]; /* reduce */
+  a2[4] ^= a2[1];
+  a2[5] ^= a2[1];
+
+  r[0] ^= a2[1] & b[7]; /* add */
+  r[1] ^= a2[2] & b[7];
+  r[2] ^= a2[3] & b[7];
+  r[3] ^= a2[4] & b[7];
+  r[4] ^= a2[5] & b[7];
+  r[5] ^= a2[6] & b[7];
+  r[6] ^= a2[7] & b[7];
+  r[7] ^= a2[0] & b[7];
+}
+
+/*
+ * Square `x` in GF(2^8) and write the result to `r`. `r` and `x` may overlap.
+ */
+static void gf256_square(uint32_t r[8], const uint32_t x[8]) {
+  uint32_t r8, r10, r12, r14;
+  /* Use the Freshman's Dream rule to square the polynomial
+   * Assignments are done from 7 downto 0, because this allows the user
+   * to execute this function in-place (e.g. `gf256_square(r, r);`).
+   */
+  r14 = x[7];
+  r12 = x[6];
+  r10 = x[5];
+  r8 = x[4];
+  r[6] = x[3];
+  r[4] = x[2];
+  r[2] = x[1];
+  r[0] = x[0];
+
+  /* Reduce with  x^8 + x^4 + x^3 + x + 1 until order is less than 8 */
+  r[7] = r14; /* r[7] was 0 */
+  r[6] ^= r14;
+  r10 ^= r14;
+  /* Skip, because r13 is always 0 */
+  r[4] ^= r12;
+  r[5] = r12; /* r[5] was 0 */
+  r[7] ^= r12;
+  r8 ^= r12;
+  /* Skip, because r11 is always 0 */
+  r[2] ^= r10;
+  r[3] = r10; /* r[3] was 0 */
+  r[5] ^= r10;
+  r[6] ^= r10;
+  r[1] = r14;  /* r[1] was 0 */
+  r[2] ^= r14; /* Substitute r9 by r14 because they will always be equal*/
+  r[4] ^= r14;
+  r[5] ^= r14;
+  r[0] ^= r8;
+  r[1] ^= r8;
+  r[3] ^= r8;
+  r[4] ^= r8;
+}
+
+/*
+ * Invert `x` in GF(2^8) and write the result to `r`
+ */
+static void gf256_inv(uint32_t r[8], uint32_t x[8]) {
+  uint32_t y[8], z[8];
+
+  gf256_square(y, x);  // y = x^2
+  gf256_square(y, y);  // y = x^4
+  gf256_square(r, y);  // r = x^8
+  gf256_mul(z, r, x);  // z = x^9
+  gf256_square(r, r);  // r = x^16
+  gf256_mul(r, r, z);  // r = x^25
+  gf256_square(r, r);  // r = x^50
+  gf256_square(z, r);  // z = x^100
+  gf256_square(z, z);  // z = x^200
+  gf256_mul(r, r, z);  // r = x^250
+  gf256_mul(r, r, y);  // r = x^254
+}
+
+bool shamir_interpolate(uint8_t *result, uint8_t result_index,
+                        const uint8_t *share_indices,
+                        const uint8_t **share_values, uint8_t share_count,
+                        size_t len) {
+  size_t i, j;
+  uint32_t x[8];
+  uint32_t xs[share_count][8];
+  uint32_t ys[share_count][8];
+  uint32_t num[8] = {~0}; /* num is the numerator (=1) */
+  uint32_t denom[8];
+  uint32_t tmp[8];
+  uint32_t secret[8] = {0};
+  bool ret = true;
+
+  if (len > SHAMIR_MAX_LEN) return false;
+
+  /* The code below assumes that none of the share_indices are equal to
+   * result_index. We need to treat that as a special case. */
```

Comment:
If two or more `share_indices` are equal to `result_index`, the following code returns `true`. I would expect it to return `false`.

Reply:
Resolved by https://github.com/trezor/trezor-firmware/pull/85/commits/223868f552404c9f1df72691ead9581beec169d0.